RISC-V Board

kunpai opened the pull request:
As a component of the stdlib, the cache, core, processor etc. inherit from various components already existing in the standard library, to model the HiFive Unmatched as closely as possible.

kunpai committed:
stdlib: Added components for a prebuilt RISC-V board

Modelling a RISC-V board whose specs are mentioned in README.md
Modeled a cache hierarchy and CPU based on the specs.

kunpai committed:
stdlib: Added components for a prebuilt RISC-V board

Modelling a RISC-V board whose specs are mentioned in README.md
Modeled a cache hierarchy and CPU based on the specs.

Commit:
stdlib: Add interface for accessing U74 CPU in stdlib

Add enum for U74 in cpu_types. Add map from U74 enum to string. Add
condition to incorporate U74 into the board from SimpleProcessor.

kunpai committed:
Merge branch 'REU2022/riscv-validation' of https://github.com/darchr/gem5 into REU2022/riscv-validation

Commit:
stdlib: Add board and config script for HiFive Unmatched

Add board HiFiveUnmatchedBoard and run script. The run script can be
given a binary to run with argument --riscv_binary and one optional
command line argument to the binary with argument --argv.

Commit:
stdlib: Fix import for hifive_board

Correct the import for the HiFiveUnmatchedBoard in hifive-run.py.

Change-Id: I98062dd05fce1e7b5da933ecd8c124bec3da4965

kunpai committed:
stdlib: Added line for HiFive CPU

kunpai committed:
Merge branch 'REU2022/riscv-validation' of https://github.com/darchr/gem5 into REU2022/riscv-validation

Commit:
arch-riscv, cpu: Set up RiscvU74CPU simobject

This simobject is defined in src/arch/riscv/RiscvCPU.py. This
simobject is accessible in the stdlib via enum CustomCPUTypes.U74.
The simobject was implemented because currently there is no mechanism
for modifying the MinorCPU simobject via the stdlib class
AbstractProcessor.

Change-Id: Iac253cd37001df41d01c16c8d49421356b5b5c44

kunpai committed:
Zhantong's work on se_binary_workload

kunpai committed:
stdlib: Initial commit to initialize U74CPU

Inheriting from AbstractCore and using MinorCPU.

kunpai committed:
Minor changes

kunpai committed:
stdlib: Processor for U74

Inherits from AbstractProcessor

Commit:
stdlib: Make U74Processor work with HiFiveUnmatchedBoard

Instantiate simobject RiscvMinorCPU in U74Core. Set _cpu_type to MINOR
in U74Procesor. Use the new U74Processor in HiFiveUnmatchedBoard.
Fix import of U74Core in hifive_proc.py.

Change-Id: Ie2a0f307227d72ee557d14f48b747c53f98e0192

Commit:
stdlib: Explicitly define RISCV as the ISA used U74

The U74 CPU implements the RISCV ISA. Therefore, U74Core and
U74Processor should use ISA.RISCV explicitly, rather than take in the
ISA as input.

Change-Id: I82a272a124c095682752f98d0a6ce71ada8399f6

Commit:
stdlib: Remove dead code in U74Processor

U74Processor is always of type CPUTypes.MINOR. Therefore the memory
mode for the board will always be MemMode.TIMING. Also removed
parameter documention of __init__ because the only parameter is self.

Change-Id: I89f17d07ff36b628aa5bd89b9b659577a2cc257e

kunpai committed:
Formatting changes

Commit:
stdlib,arch-riscv: Restore certain stdlib and arch files

The original implementation of the U74CPU required creating a new
simobject in riscv, and a new CPUType in the stdlib. Since a proper
workaround has been implemented, the files affected by the original
implementation can be restored to their original state. These files
are restored to commit 23b2561bf668d42b51b36ed2084a499ab46996b3, which
was the head of the develop branch of darchr/gem5 at the time this
branch was created.

Change-Id: I55fec97a6ff85c79fc4a4e166a0a07f51c012a65

Commit:
stdlib: Update HiFive Unmatched README

Change-Id: I22d4aa3fba227b977b7ad3b95eca81b6611cb007

powerjg commented on src/python/gem5/components/boards/se_binary_workload.py:
```diff
@@ -23,52 +23,49 @@
 # THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 # (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-
```
I suggest not committing whitespace-only changes unless you're updating the style to be more compliant.

powerjg commented on src/python/gem5/prebuilt/hifiveunmatched/README.md:
```diff
@@ -0,0 +1,52 @@
+- L1D Associativity : 8
+- L1I Size : 32kB
+- L1I Associativity : 4
+- L2 Size : 2MB
```
Note (in theory) this is actually configurable. We may want to make this a parameter of the board.

powerjg commented on src/python/gem5/prebuilt/hifiveunmatched/hifive-run.py:
```diff
@@ -0,0 +1,59 @@
+# Copyright (c) 2022 The Regents of the University of California
```
The "runscript" shouldn't be in `src/`. It should be in `configs/examples` or `configs/known_good`.

jasrangi replied:
The location is now configs/example/gem5_library/

powerjg commented on src/python/gem5/prebuilt/hifiveunmatched/hifive-run.py:
```diff
@@ -0,0 +1,59 @@
+
+# collect optional CLI arg for RISCV binary to run
+parser = argparse.ArgumentParser(description="Binary to run on system")
+parser.add_argument(
+    "--riscv_binary", type=str, help="The RISCV binary to execute on the CPU",
+    default="riscv-hello",
+)
```
A required argument shouldn't be `--<something>`. Also, an argument that the user *must* consider (e.g., what workload they want to run) shouldn't have a default.

jasrangi replied:
Removed the default and set it to "riscv_binary" instead of "--riscv_binary".

powerjg commented on src/python/gem5/prebuilt/hifiveunmatched/hifive-run.py:
```diff
@@ -0,0 +1,59 @@
+else:
+    board.set_se_binary_workload(
+        CustomResource(args.riscv_binary),
+        arguments=[args.argv]
```
You can remove the `if`/`else` if you use `args.argv.split()`

jasrangi replied:
The `if`/`else` exists to use Resource() when riscv-hello must be run, and use CustomResource() otherwise. Is there a better way to do that?
Also I pushed a change that uses split() to allow for multiple arguments in case future benchmarks require more than one.

powerjg replied:
I think that it should be a CustomResource in all cases. You the argument is the path to a binary, not the name of a resource.

It would be confusing if you could use the same argument to specify either a resource (of which there is only one? What happens when there's more?) or a path.

jasrangi replied:
Ok that makes sense. It is now using CustomResource() only.

powerjg commented on src/python/gem5/prebuilt/hifiveunmatched/hifive-run.py:
```diff
@@ -0,0 +1,59 @@
+    )
+
+simulator = Simulator(board=board)
+simulator.run()
```
One newline at the end of the file is nice.

powerjg commented on src/python/gem5/prebuilt/hifiveunmatched/hifive-run.py:
```diff
@@ -0,0 +1,59 @@
+)
+args = parser.parse_args()
+
+board = HiFiveUnmatchedBoard()
```
I love how simple this is!

powerjg commented on src/python/gem5/prebuilt/hifiveunmatched/hifive_board.py:
```diff
@@ -0,0 +1,56 @@
+    def __init__(self) -> None:
+        requires(isa_required=ISA.RISCV)
+
+        cache_hierarchy = HiFiveCacheHierarchy(
+            l1d_size="32kB", l1i_size="32kB", l2_size="2MB"
+        )
```
Perfect!

powerjg commented on src/python/gem5/prebuilt/hifiveunmatched/hifive_board.py:
```diff
@@ -0,0 +1,56 @@
+        memory.set_memory_range(
+            [AddrRange(start=0x80000000, size=memory.get_size())])
+
+        processor = U74Processor()
```
Wonderful!

powerjg commented on src/python/gem5/prebuilt/hifiveunmatched/hifive_cache.py:
```diff
@@ -0,0 +1,185 @@
+            for i in range(board.get_processor().get_num_cores())
+        ]
+        # ITLB Page walk caches
+        self.iptw_caches = [
+            MMUCache(size="4KiB")
+            for _ in range(board.get_processor().get_num_cores())
+        ]
+        # DTLB Page walk caches
+        self.dptw_caches = [
+            MMUCache(size="4KiB")
+            for _ in range(board.get_processor().get_num_cores())
+        ]
```
Out of curiosity, are these listed in the spec?

kunpai replied:
Yeah, I might be mistaken in interpreting it, but on page 55 (section 4.7.7) of the datasheet, there is a mention of the level 2 TLB pagesize being 4 KiB. On writing this, I realize the importance of citing the page number and section from where we picked up the data :)

powerjg replied:
A two level TLB and a PWC are two different things. The current implementation is good with me, but we should make a note of the discrepancy.

powerjg commented on src/python/gem5/prebuilt/hifiveunmatched/hifive_cache.py:
```diff
@@ -0,0 +1,185 @@
+class HiFiveCacheHierarchy(
+    AbstractClassicCacheHierarchy, AbstractTwoLevelCacheHierarchy
+):
+    """
+    A cache setup where each core has a private L1 Data and Instruction Cache,
+    and a private L2 cache.
+    """
```
Point to where the information is for this hierarchy in the spec.

powerjg commented on src/python/gem5/prebuilt/hifiveunmatched/hifive_cache.py:
```diff
@@ -0,0 +1,185 @@
+
+    def __init__(
+        self,
+        l1d_size: str,
+        l1i_size: str,
+        l2_size: str,
+        membus: BaseXBar = _get_default_membus.__func__(),
```
Are these parameters needed?

I would argue that there should be just one parameter, which is the programmable size of the L2 (number of ways??? I can't remember how it's programmed specifically).

powerjg commented on src/python/gem5/prebuilt/hifiveunmatched/hifive_core.py:
```diff
@@ -0,0 +1,154 @@
+            executeFuncUnits = U74FUPool(),
+            branchPred = U74BP())
+        self.core.createThreads()
+    def get_simobject(self) -> BaseCPU:
+        return self.core
+    @overrides(AbstractCore)
+    def get_isa(self) -> ISA:
+        return self._isa
+    @overrides(AbstractCore)
+    def connect_icache(self, port: Port) -> None:
+        self.core.icache_port = port
+    @overrides(AbstractCore)
+    def connect_dcache(self, port: Port) -> None:
+        self.core.dcache_port = port
+    @overrides(AbstractCore)
+    def connect_walker_ports(self, port1: Port, port2: Port) -> None:
+        self.core.mmu.connectWalkerPorts(port1, port2)
+    @overrides(AbstractCore)
+    def set_workload(self, process: Process) -> None:
+        self.core.workload = process
+    @overrides(AbstractCore)
+    def set_switched_out(self, value: bool) -> None:
+        self.core.switched_out = value
+    @overrides(AbstractCore)
+    def connect_interrupt(
+        self, interrupt_requestor: Optional[Port] = None,
+        interrupt_responce: Optional[Port] = None
+    ) -> None:
+        # TODO: This model assumes that we will only create an interrupt
+        # controller as we require it. Not sure how true this is in all cases.
+        self.core.createInterruptController()
+    @overrides(AbstractCore)
+    def get_mmu(self) -> BaseMMU:
+        return self.core.mmu
```
I suggest adding some whitespace :)

powerjg commented on src/python/gem5/prebuilt/hifiveunmatched/hifive_proc.py:
```diff
@@ -0,0 +1,64 @@
+    def _create_cores(
+        self,
+    ):
+        return U74Core()
```
Shouldn't this be `[U74Core() for i in range(4)]`?

kunpai replied:
Hi Jason! Thanks for the feedback. I believe the MinorCPU in stdlib only runs one core. When we tried to set the number of cores as 4, it panicked. @jasrangi can you confirm that I recall correctly?

powerjg replied:
You'll need to set them all up correctly 😆.

I guess 1 core is OK for now. But we need to be able to support the same configuration at some point.

kunpai replied:
Oh I see, yeah I agree with you. We might need to set it up later. For now, even when we run with ./perf we taskset the workload so that there is no migration of CPU cores

powerjg commented on src/python/gem5/prebuilt/hifiveunmatched/hifive_proc.py:
```diff
@@ -0,0 +1,64 @@
+            cores=self._create_cores()
+        )
+
+    def _create_cores(
+        self,
+    ):
```
No need for the newlines here.

powerjg reviewed (commented):
This looks *really* great! I'm excited to see the data!

Commit:
stdlib: Move HiFive run script to configs directory

hifive-run.py should be in configs/, not src/. Its new location is
configs/example/gem5_library.

Change-Id: I9f52174454dc28540d1b6d585e89dbc18a033315

Commit:
stdlib: Update HiFive Unmatched board README

Update the README to show the new location of the run script.

Change-Id: I2cfec2ce13911d83593be6d549832b274db51ff4

Commit:
stdlib: Fix imports in hifive run script

Moving the hifive-run.py script into config/ caused multi-level import
issues. Added necessary files to src/python/SConscript. Rewrote the
includes.

Change-Id: Ifa0943a5056d561010b401cd78fa51822c9bb3f7

kunpai committed:
stdlib: Formatting changes

kunpai committed:
stdlib: Ran black on all files

Commit:
stdlib: Set riscv_binary as a required arg for hifive-run.py

For the hifive run script, the riscv binary to run must be a required
argument.

Change-Id: Ic86222e4597758a54a70e923bea00e46afb597e6

Commit:
stdlib: Allow multiple argv in hifive-run.py

In the hifive-run.py script, allow multiple arguments to be passed to
the RISCV binary that will be run. An example is passing flag
--argv="100 123123 12", which will pass argv array
['100', '123123', '12] to the binary.

Change-Id: Ic0a217c58bc3000c38d30adb7d76af50c3949439

Commit:
stdlib: Always use CustomResource in hifive-run script

Change-Id: Ie8755c6164ecc32d56a6ac74cf20d80e54245130

kunpai committed:
stdlib: Included params for assoc on caches

kunpai committed:
Merge branch 'REU2022/riscv-validation' of https://github.com/darchr/gem5 into REU2022/riscv-validation

kunpai commented:
src/python/gem5/components/boards/se_binary_workload.py is a file we had Zhantong provide to us based on her work, how do I remove it from this PR?

kunpai committed:
stdlib: Removed params for assoc and L1 sizes on caches

kunpai committed:
stdlib: Removed params for assoc and L1 sizes on caches

powerjg commented on src/python/gem5/prebuilt/hifiveunmatched/hifive_cache.py:
```diff
@@ -0,0 +1,175 @@
+    """
+
+    @staticmethod
+    def _get_default_membus() -> SystemXBar:
+        """
+        A method used to obtain the default memory bus of 64 bit in width for
+        the PrivateL1PrivateL2 CacheHierarchy.
+        :returns: The default memory bus for the PrivateL1PrivateL2
+        CacheHierarchy.
+        :rtype: SystemXBar
+        """
+        membus = SystemXBar(width=64)
+        membus.badaddr_responder = BadAddr()
+        membus.default = membus.badaddr_responder.pio
+        return membus
```
I don't think this is necessary. We don't need this "option" in the HiFive-specific board.

powerjg commented on src/python/gem5/prebuilt/hifiveunmatched/hifive_cache.py:
```diff
@@ -0,0 +1,175 @@
+    def __init__(
+        self,
+        l2_size: str,
+        membus: BaseXBar = _get_default_membus.__func__(),
```
Ditto. I don't think this parameter is needed.

powerjg commented on src/python/gem5/prebuilt/hifiveunmatched/hifive_cache.py:
```diff
@@ -0,0 +1,175 @@
+    AbstractClassicCacheHierarchy,
+)
+from gem5.components.cachehierarchies.abstract_two_level_cache_hierarchy import (
+    AbstractTwoLevelCacheHierarchy,
```
@BobbyRBruce, can you think about whether it makes sense to inherit from this class?

powerjg reviewed (commented):
Overall, it looks good.

One big thing that's missing is support for full system mode. Let's start focusing on that so we can get some more realistic results.

kunpai committed:
stdlib: Removed param for membus and made it default instead

kunpai committed:
stdlib: Init for FS mode and other changes

Added a function that returns memory in hifive_board
Changed Processor, Cache and Mem in FS board to use SE components.

kunpai committed:
stdlib: Whitespaces to improve readability

Commit:
stdlib: Initialize HiFive FS board with components

Change-Id: Id38f2e0d2320822bf2a6af8eb5c83f3cc9a7ec78

kunpai committed:
stdlib: Changes to RTC to match datasheet params

kunpai committed:
stdlib: Added documention for cache, core

Added documentation to where the data for the choices we made
for the board exists.

kunpai committed:
stdlib: Added FS mode

HiFive Board can run both FS and SE mode depending on a param passed into it.

Commit:
stdlib: Remove hifive_fs board

hifive_board.py has been updated to handle both SE and FS mode. Remove
hifive_fs.py.

Change-Id: Iec3f728f7841d45fc692d3dc50cebf36d1cbe808

Commit:
stdlib: Update hifive runscript to handle FS and SE mode

hifive-run.py sets FS or SE workload depending on boolean argument
"--fullsystem". The default value is "False", so the runscript will
run the board in SE mode by default. Also, the memory function is
defined in hifive_board.py for the board's memory setup.

Change-Id: I1f70a8713d090674c4fea174e295719ef179669d

kunpai committed:
stdlib: Added documentation for memory

Commit:
stdlib: Remove unused imports in hifiveunmatched board

Change-Id: I30d27a6a2c6ddb0066df5a6228a67c2e8a6f669e

kunpai committed:
removing a file from PR

kunpai committed:
stdlib: Changed params in exec pipeline to ARM Research

Commit:
stdlib: Use core IDs in hifiveunmatched cores

Use core IDs in the U74Core to allow for use of multiple cores.
FS mode with one core boots successfully to the login screen. Only
one core has been tested to work with riscv-disk-img at this time.

Change-Id: I09c424db3248c31a8811c63180be2ca03a5b32d0

kunpai committed:
stdlib: Replaced ChanneledMemory with DDR4

kunpai committed:
Merge branch 'REU2022/riscv-validation' of https://github.com/darchr/gem5 into REU2022/riscv-validation

kunpai committed:
stdlib: Increased decodeToExecuteForwardDelay to avoid PMC fault

kunpai committed:
stdlib: Increased decodeToExecuteForwardDelay to avoid PMC fault

powerjg commented on src/python/gem5/prebuilt/hifiveunmatched/hifive_board.py:
```diff
@@ -0,0 +1,579 @@
+            self.platform.clint.num_threads = self.processor.get_num_cores()
+
+            # Add the RTC
+            """
+            The frequency of the RTC for the system is set to 1MHz. 
+            Details can be found on page 77, section 7.1 of the datasheet.
+
+            """
```
`"""` should only be used for pydoc statements, not for general comments

Commit:
stdlib: Use 4 cores in FS mode for HiFiveUnmatched board

If the HiFiveBoard object is created in FS mode, 4 cores will be
generated by _create_cores(). FS mode with 4 cores has been tested to
work. It took 2.5 hours real time to fully boot and reach the login
screen on Amarillo. The nproc command and /proc/cpuinfo confirmed that
there are 4 cores. The cpuinfo was nearly identical to the hardware
HiFiveUnmatched, with the  exception being SV39 MMU for hardware
and SV48 MMU for the gem5 board.

Change-Id: I7d4bd74cfbd5c12043ebb7feaebba34fc873901b

kunpai committed:
stdlib: Changed documentation to include RTCin the intro

kunpai committed:
stdlib: Changed documentation to include RTCin the intro

Commit:
stdlib: Pass cache parameters to caches in HiFive Unmatched

Pass the cache associativity parameters to L1I, L1D, and L2 caches.
Prior, these parameters were set in the AbstractTwoLevelCacheHierarchy
parent class but were not passed to the __init__ functions for the
caches. The associativities are now manually set to 4, 8, and 16 for
L1I, L1D, and L2 caches respectively, as defined by the official
manual for the board. Prior, the defaults were used, which were 8, 8,
and 16 respectively at this time.

Change-Id: Ic5ddca9fe4c938bfac83cc2a09baa0ff4898121c

Commit:
cpu-minor: Trying branch stats on Minor CPU

Change-Id: I79afd3a53cd81ff75de67261efab845549660af5

kunpai committed:
Merge branch 'REU2022/riscv-validation' of https://github.com/darchr/gem5 into REU2022/riscv-validation

kunpai committed:
cpu-minor: Completed implementation of new Control Stats

Commit:
cpu-minor: Add control instruction stats to Minor CPU

Add control/branch instructions committed stat to Minor CPU. The stats
can be found in board.processor.cores.core.commitedControl_0 in
stats.txt. The stats counted are Call, Return, DirectCtrl,
IndirectCtrl, CondCtrl, UncondCtrl, and Control. The last one tracks
the total control/branch instructions committed.

Change-Id: I0f2fa9540a1002d6ef73be7e1ef677f4b27eaf83

powerjg commented on src/cpu/minor/execute.cc:
```diff
@@ -884,6 +884,47 @@ Execute::doInstCommitAccounting(MinorDynInstPtr inst)
+    if (inst->staticInst->isControl())
+    {
+        if (inst->staticInst->isReturn())
+        {
```
```suggestion
    if (inst->staticInst->isControl()) {
        if (inst->staticInst->isReturn()) {
```
brackets should be on the same line.

powerjg commented on src/cpu/minor/stats.cc:
```diff
@@ -79,6 +95,10 @@ MinorStats::MinorStats(BaseCPU *base_cpu)
     committedInstType.ysubnames(enums::OpClassStrings);
+
+    commitedControl
+        .init(base_cpu->numThreads, enums::ControlInst::NumControlInstTypes);
```
Can you use `StaticInstFlags` instead?

kunpai replied:
Yeah, sure, but these functions are predefined and return the StaticInstFlags I believe. Just so I understand gem5 better, why would we not use this function?

powerjg replied:
I was suggesting this to avoid creating a new enum.

jasrangi replied:
Enum StaticInstFlags.Flags of build/RISCV/enums/StaticInstFlags.hh has 36 different flags. From my understanding, if we use this our statistics vector will be of length 36 (currently it is 7). Would it be better to use those flags and extend the statistics to track all 36 StaticInst types?

powerjg replied:
I think it is better to extend the stat vector to be length 36, but only update the stats `if (isControl())`. You can set the option on the statistic so that it only displays if the value is non-zero to hide all of the others.

kunpai replied:
So use .flags(statistics::dist) I believe, right?

powerjg replied:
I don't think so. See https://www.gem5.org/documentation/general_docs/statistics/ you can use the `nozero` flag.

kunpai replied:
Oh yeah you're right, I was reading the wrong comment in src/base/stats/info.hh lol

jasrangi replied:
Removed the old enum and string map and used StaticInstFlags. Commit [315ae16](https://github.com/darchr/gem5/pull/247/commits/315ae16cbfad64f56622f14cb9437ec2f958fbe7).

powerjg commented on src/python/gem5/prebuilt/hifiveunmatched/hifive_board.py:
```diff
@@ -0,0 +1,577 @@
+            kernel_args=kernel_args,
+            exit_on_work_items=exit_on_work_items,
+        )
+        
```
Something is wrong with this whitespace :)

kunpai committed:
Update src/cpu/minor/execute.cc

Co-authored-by: Jason Lowe-Power <jason@lowepower.com>

Commit:
cpu-minor: Fix typo in committedControl stats vector name

Change-Id: I5048b4e84c4a4c62ec502f7d5fe630e6e05e96b2

kunpai committed:
cpu-minor: Added nozero flag to control instructions

Commit:
cpu-minor: Use StaticInstFlags for tracking Control stats

StaticInstFlags::Flags enum is a superset of enum::ControlInst. Using
struct StaticInstFlags allows tracking of control instructions
committed without creating a new enum. The CommittedControl vector
has space to track stats for all flags but will only track control
instructions. The nozero flag is used to not output anything in
stats.txt for the non-control flags in StaticInstFlags::Flags.

Change-Id: Iad35ef0ee2a8f2674beafcbd85404c3cff45cf46

Commit:
stdlib: Add documentation for U74CPU pipeline parameters

Change-Id: I947a2a3614404ebeaedfb89f1e7a466c6ae8db60

Commit:
stdlib: Pass disk image as argument to hifive-run config script

Prior, the default disk image used in fullsystem mode for the hifive
config script was riscv-disk-img. Now the name of a provided gem5
resource can be passed as the second positional argument to the run
script. Paths to custom disk images can also be passed.

Change-Id: I287ef31c18f3a333ab71d0bf3929cf3734cb45f9
